File: 2022/day03/go/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
	"time"
)

func p1(data []string, priorities_str string) int {
	var sum int = 0
	for _, rucksack := range data {
		var l = len(rucksack) / 2
		var c1 = rucksack[:l]
		var c2 = rucksack[l:]

		for i := 0; i < len(c1); i++ {
			char := string(c1[i])
			exists := strings.Contains(c2, char)

			if exists {
				idx := strings.Index(priorities_str, char)
				sum += idx + 1
				break
			}
		}
	}

	return sum
}

func p2(data []string, priorities_str string) int {
	var sum int = 0
	var counter int = 0
	var history = [3]string{}

	for _, item := range data {
		history[counter] = item
		counter += 1

		if counter == 3 {
			counter = 0

			for i := 0; i < len(history[0]); i++ {
				c1 := string(history[0][i])

				exists_in_2 := strings.Contains(history[1], c1)

				if exists_in_2 {
					exists_in_3 := strings.Contains(history[2], c1)

					if exists_in_3 {
						idx := strings.Index(priorities_str, c1)
						sum += (idx + 1)
						break
					}
				}
			}
		}
	}

	return sum
}

func main() {
	start := time.Now()

	priorities_str := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"

	// load the data into a slice of strings
	bs, err := ioutil.ReadFile("../input.txt")

	if err != nil {
		bs, _ = ioutil.ReadFile("../input.txt")
	}

	input := strings.Split(string(bs), "\n")

	fmt.Println("The solution to part 1 is: ", p1(input, priorities_str))
	fmt.Println("The solution to part 2 is: ", p2(input, priorities_str))

	end := time.Since(start)
	fmt.Println("Time: ", end)

}

// go run ../go/main.go

File: 2022/day03/nim/main.nim
import std/strutils
import std/[times, sets]
import sequtils


proc p1(data: seq[string], priorities: seq[char]): int =
    var sum: int = 0

    for item in data:
        var l: int = len(item) div 2
        var c1: string = item[0..l-1]
        var c2: string = item[l..^1]

        var a1 = toHashSet(toSeq(c1.items))
        var a2 = toHashSet(toSeq(c2.items))
        var i = intersection(a1, a2).toSeq[0]
        sum += (priorities.find(i) + 1)

    return sum


proc p2(data: seq[string], priorities: seq[char]): int =
    var sum:int = 0
    var counter: int = 0
    var history: array[3, string]

    for item in data:
        history[counter] = item
        counter += 1

        if counter == 3:
            var a1 = toHashSet(toSeq(history[0].items))
            var a2 = toHashSet(toSeq(history[1].items))
            var a3 = toHashSet(toSeq(history[2].items))

            var i1 = intersection(a1, a2)
            var i2 = intersection(i1, a3)
            var i3 = intersection(i1, i2)

            var i: char = i3.toSeq()[0]
            sum += (priorities.find(i) + 1)

            counter = 0

    return sum


let time = cpuTime()

let data = readFile("../input.txt").splitLines()

var priorities: seq[char] = @[]
for c in 'a'..'z':
    priorities.add(c)

for c in 'A'..'Z':
    priorities.add(c)    


echo "The solution to part 1 is: ", p1(data, priorities)
echo "The solution to part 2 is: ", p2(data, priorities)

echo "Time taken: ", (cpuTime() - time) * 1000, "ms"

# nim c -r -d:release -d:danger main.nim